// pdf-creator-node/dist/pdfChrome.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildPdfChrome = buildPdfChrome;
exports.mergePdfCreateOptions = mergePdfCreateOptions;
function escapeHtml(s) {
    return s
        .replace(/&/g, "&amp;")
        .replace(/</g, "&lt;")
        .replace(/>/g, "&gt;")
        .replace(/"/g, "&quot;");
}
function defaultHeaderHtml(title) {
    return `<div style="text-align:center;font-size:10px;padding:6px;">${escapeHtml(title)}</div>`;
}
function buildFooterContents(copyright, showPageNumbers) {
    const hasCopyright = copyright != null && String(copyright).trim() !== "";
    const pagePart = showPageNumbers
        ? '<span style="color:#444">{{page}}</span>/<span>{{pages}}</span>'
        : "";
    if (hasCopyright && pagePart) {
        return {
            default: `<div style="text-align:center;font-size:9px;padding:4px;">${escapeHtml(String(copyright))} · ${pagePart}</div>`,
        };
    }
    if (hasCopyright) {
        return {
            default: `<div style="text-align:center;font-size:9px;padding:4px;">${escapeHtml(String(copyright))}</div>`,
        };
    }
    if (pagePart) {
        return {
            default: `<div style="text-align:center;font-size:9px;padding:4px;">${pagePart}</div>`,
        };
    }
    return { default: "" };
}
/**
 * Builds partial html-pdf options from layout / header / footer / copyright.
 * Use {@link mergePdfCreateOptions} to merge with explicit `format`, `header`, `footer`.
 */
function buildPdfChrome(chrome) {
    var _a, _b, _c, _d;
    if (!chrome)
        return {};
    const out = {};
    if (chrome.layout) {
        const L = chrome.layout;
        if (L.format)
            out.format = L.format;
        if (L.orientation)
            out.orientation = L.orientation;
        if (L.width)
            out.width = L.width;
        if (L.height)
            out.height = L.height;
        if (L.border !== undefined)
            out.border = L.border;
    }
    if (chrome.header) {
        const h = chrome.header;
        const contents = (_a = h.html) !== null && _a !== void 0 ? _a : (h.title != null && h.title !== "" ? defaultHeaderHtml(h.title) : undefined);
        if (contents) {
            out.header = {
                height: (_b = h.height) !== null && _b !== void 0 ? _b : "45mm",
                contents,
            };
        }
        else if (h.height) {
            out.header = { height: h.height };
        }
    }
    if (chrome.footer) {
        const f = chrome.footer;
        if (f.html) {
            out.footer = {
                height: (_c = f.height) !== null && _c !== void 0 ? _c : "28mm",
                contents: { default: f.html },
            };
        }
        else {
            const hasCopyright = f.copyright != null && String(f.copyright).trim() !== "";
            const showPages = f.showPageNumbers === true ||
                (f.showPageNumbers !== false && hasCopyright);
            if (hasCopyright || showPages) {
                const contents = buildFooterContents(hasCopyright ? String(f.copyright) : undefined, showPages);
                out.footer = {
                    height: (_d = f.height) !== null && _d !== void 0 ? _d : "28mm",
                    contents,
                };
            }
            else if (f.height) {
                out.footer = { height: f.height };
            }
        }
    }
    return out;
}
/**
 * Merges `pdfChrome` with explicit html-pdf fields. Explicit `header` / `footer` / `format` win.
 */
function mergePdfCreateOptions(options) {
    var _a, _b;
    const { handlebarsHelpers: _h, pdfChrome } = options, userPdfOptions = __rest(options, ["handlebarsHelpers", "pdfChrome"]);
    const fromChrome = buildPdfChrome(pdfChrome);
    return Object.assign(Object.assign(Object.assign({}, fromChrome), userPdfOptions), { header: (_a = userPdfOptions.header) !== null && _a !== void 0 ? _a : fromChrome.header, footer: (_b = userPdfOptions.footer) !== null && _b !== void 0 ? _b : fromChrome.footer });
}
